doc/developer/temporal-filters.md

```diff
@@ -0,0 +1,67 @@
+# Temporal Filters
+
+Materialize is able to introduce and remove records as a function of timestamps presented as data in the records.
+This is performed using idiomatic SQL, with some restrictions, but it can be hard to understand at first.
+
+Consider the following SQL query:
+```sql
+SELECT *
+FROM data
+WHERE mz_logical_timestamp() BETWEEN data.valid_from AND data.valid_until;
+```
+The `WHERE` clause here relates two columns in the data to `mz_logical_timestamp()`, a function that returns the current logical timestamp.
+For one-shot queries `mz_logical_timestamp()` is simply evaluated and replaced by that timestamp when the query executes.
+In a *maintained* query, as for example when you call `CREATE MATERIALIZED VIEW` or `CREATE INDEX`, the value of the function changes every millisecond.
+In particular, in a maintained query any row will not be present at timestamps before its `valid_from` column nor after `valid_until` column.
+
+Materialize does not re-evaluate the predicate every millisecond for every row.
+Understanding what Materialize does do may help you understand the behavior of and restrictions on temporal filters.
+
+## Manipulating timestamped updates
+
+Materialize records *changes* to collections of data.
+Each change occurs at some "logical timestamp", an explicit transcription of the time at which a change should occur.
+Generally, an update occurs at its logical timestamp and is then in effect indefinitely.
+To undo an update one issues a new update with a negated sign, at a future logical timestamp.
+
+We can restrict the duration of an update, to any non-empty interval `[from, until]`, by
+1. advancing the update's timestamp to at least `from`, and
+2. introducing a new negated update at timestamp `until`.
+The first step ensures that the update does not happen until `from` and the second step ensures that the update is retracted at `until`.
+Each of the two steps is optional, if either of the `from` or `until` bounds are unspecified.
+
+This mechanism allows us to support `WHERE` clauses in which `mz_logical_timestamp()` is directly related to an expression not containing `mz_logical_timestamp()`, or conjunctions (`AND`) of these clauses.
+The following query demonstrates several example uses:
+```sql
+SELECT *
+FROM data
+WHERE mz_logical_timestamp() BETWEEN data.valid_from AND data.valid_until
+  AND mz_logical_timestamp() > 100
+  AND 567 > mz_logical_timestamp();
+```
+The `BETWEEN` expression simplifies to two inequalities, and the other two inequalities are one-sided bounds.
+The `mz_logical_timestamp()` call can appear on either side of the inequality, but it must occur by itself.
+The inequality cannot be "not equal" (`!=` or `<>`) as this results in a "hole" rather than an upper and lower bound.
```

Comment:
I think it might be nice to have an example of a query with a one-sided bound. The text makes it clear enough, but all the examples (and the wording of #5624) only focus on two-sided bounds.

Reply:
Can do. The above was meant to show off that there are a bunch of weird ways you can do things, but you are right that it does not clarify that you can do any subset of them too.